status/api/provider.nim

```diff
@@ -23,3 +28,59 @@ proc callRpc*(self: StatusObject, rpcMethod: string, params: JsonNode):
     return CallRpcResult.ok await self.web3.callRpc(rpcMethod, params)
   except CatchableError as e:
     return CallRpcResult.err e.msg
+
+proc signTransaction(tr: var Transaction, pk: PrivateKey) =
+  let h = tr.txHashNoSignature
+  let s = sign(pk, SkMessage(h.data))
+
+  var r = toRaw(s)
+  let v = r[64]
+
+  tr.R = fromBytesBE(UInt256, r.toOpenArray(0, 31))
+  tr.S = fromBytesBE(UInt256, r.toOpenArray(32, 63))
+
+  tr.V = int64(v)
+
+  if tr.txType == TxType.TxLegacy:
+    tr.V += 27 # TODO! Complete this
+
+proc sendTransaction*(self: StatusObject, fromAddress: EthAddress, transaction: Transaction, password: string, dir: string):
+  Future[SendTransactionResult] {.async.} =
+  if not self.isLoggedIn:
+    return SendTransactionResult.err "Not logged in. Must be logged in"
+
+  try:
+    let accounts = self.userDb.getWalletAccounts()
+    let account = accounts.filter(x => x.address == Address(fromAddress))
+    if account.len == 0:
+      return SendTransactionResult.err "Wallet address not found"
+
+    let loadAccountResult = self.accountsGenerator.loadAccount(account[0].address, password,dir)
```

Comment:
It's probably a matter of my being ignorant re: how some of the crypto parts work, but is loading the keystore from disk every time simply how we have to do it? Does status-go do the same?

Or is it a matter of something/s re: nim-status' sqlcipher database and related procs not being wired up in a way that would allow us to use info from the open database instead of loading the keystore?

Reply:
I was looking at the database and it seems that the private keys are not stored. This is the reason why I'm loading the keystore from disk instead of obtaining the info from the DB.